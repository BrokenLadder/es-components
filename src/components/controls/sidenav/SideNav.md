```
<div style={{display: 'flex'}}>
  <div style={{width: '30%', marginRight: '4em'}}>
    <h3>Standard</h3>
    <SideNav defaultSelected="home">
      <SideNav.Item id="home"><a>Home</a></SideNav.Item>
      <SideNav.Item id="cart">Cart</SideNav.Item>
      <SideNav.Item id="help">Help</SideNav.Item>
      <SideNav.Item id="info"><a>About</a></SideNav.Item>
      <SideNav.Item id="info" isDisabled>Disabled</SideNav.Item>
    </SideNav>
  </div>

  <div style={{width: '30%'}}>
    <h3>Alternate Style</h3>
    <SideNav defaultSelected="home" altStyle>
      <SideNav.Item id="home"><a>Home</a></SideNav.Item>
      <SideNav.Item id="cart">Cart</SideNav.Item>
      <SideNav.Item id="help">Help</SideNav.Item>
      <SideNav.Item id="info"><a>About</a></SideNav.Item>
      <SideNav.Item id="info" isDisabled>Disabled</SideNav.Item>
    </SideNav>
  </div>
</div>
```

Setting ```onItemSelected``` will fire the function when a ```SideNav.Item``` is clicked.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" onItemSelection={ (id) => {alert(id); }}>
    <SideNav.Item id="home">Home</SideNav.Item>
    <SideNav.Item id="cart">Cart</SideNav.Item>
    <SideNav.Item id="info" isDisabled>Disabled</SideNav.Item>
  </SideNav>
</div>
```

Set ```onNavClick``` on an individual ```SideNav.Item```.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" altStyle>
    <SideNav.Item id="home" onNavClick={ () => (alert('Time to go Home!'))}>Home</SideNav.Item>
    <SideNav.Item id="cart" onNavClick={ () => (alert('You clicked the Cart!'))}>Cart</SideNav.Item>
    <SideNav.Item id="info" isDisabled>Disabled</SideNav.Item>
  </SideNav>
</div>
```

Manage selected item state manually using ```selected``` instead of ```defaultSelected```.
```
<div style={{width: '30%'}}>
  <SideNav selected={state.selected} onItemSelection={ (id) => {setState({selected: id})} }>
    <SideNav.Item id="home"><a>Home</a></SideNav.Item>
    <SideNav.Item id="cart">Cart</SideNav.Item>
    <SideNav.Item id="info"><a>Help</a></SideNav.Item>
  </SideNav>
</div>
```
