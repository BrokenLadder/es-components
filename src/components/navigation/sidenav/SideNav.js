import React, { Children, Component, cloneElement } from 'react';
import PropTypes from 'prop-types';
import { colors } from '../../theme';
import styled from 'styled-components';
import { noop } from 'lodash';

import NavItem from './NavItem';

const NavStyled = styled.nav`
  background-color: ${colors.white};
  box-shadow: 2px 2px 5px ${colors.grayDark};

  ul {
    list-style: none;
    margin: 0;
    padding: 0;
  }
`;

class SideNav extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: props.defaultSelected,
      defaultSelected: props.defaultSelected
    };
  }

  onNavClick = (navId = null) => {
    const { onItemSelected = noop } = this.props;

    if (this.state.defaultSelected) {
      this.setState({ selected: navId }, () => {
        onItemSelected(navId);
      });
    } else {
      onItemSelected(navId);
    }
  };

  render() {
    const { altStyle, children, selected } = this.props;
    return (
      <NavStyled>
        <ul>
          {Children.toArray(children).map(child => {
            if (child !== null && child.type === NavItem) {
              const currentSelected = this.state.defaultSelected
                ? this.state.selected
                : selected;
              return cloneElement(child, {
                altStyle,
                highlightedId: currentSelected,
                onClick: this.onNavClick
              });
            }
            return child;
          })}
        </ul>
      </NavStyled>
    );
  }
}

SideNav.propTypes = {
  /** Use the alternate nav style */
  altStyle: PropTypes.bool,
  children: PropTypes.node,
  /** Use to manually select nav item by id, stateless mode */
  selected: PropTypes.string,
  /** Use to set a default nav, stateful mode */
  defaultSelected: PropTypes.string,
  /** Function called when a nav item is clicked */
  onItemSelected: PropTypes.func
};

SideNav.Item = NavItem;

export default SideNav;
