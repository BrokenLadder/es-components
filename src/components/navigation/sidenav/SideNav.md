```
<div style={{display: 'flex'}}>
  <div style={{width: '30%', marginRight: '4em'}}>
    <h3>Standard</h3>
    <SideNav defaultSelected="home">
      <SideNav.Item id="home" title="Home" targetUrl="#home" />
      <SideNav.Item id="cart" title="Cart">Cart</SideNav.Item>
      <SideNav.Item id="help" title="Help" />
      <SideNav.Item id="info" title="About" />
      <SideNav.Item id="disabled" title="Diabled" isDisabled />
    </SideNav>
  </div>

  <div style={{width: '30%'}}>
    <h3>Alternate Style</h3>
    <SideNav defaultSelected="home" altStyle>
      <SideNav.Item id="home" title="Home" />
      <SideNav.Item id="cart" title="Cart">Cart</SideNav.Item>
      <SideNav.Item id="help" title="Help" />
      <SideNav.Item id="info" title="About" />
      <SideNav.Item id="disabled" title="Diabled" isDisabled />
    </SideNav>
  </div>
</div>
```

Assign a ```targetUrl``` to set the nav link href location. Use ```isExternalLink``` to open the link in a new browser window.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home">
    <SideNav.Item id="home" targetUrl="/" title="Home" onNavClick={ () => (alert('Time to go Home!'))} />
    <SideNav.Item id="medicare" targetUrl="https://medicare.oneexchange.com" title="Medicare" isExternalLink />
    <SideNav.Item id="alerts" targetUrl="#alert" title="Alert Section" />
  </SideNav>
</div>
```

Setting ```onItemSelected``` will execute the function when a ```SideNav.Item``` is clicked.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" onItemSelected={ (id) => {alert(id); }}>
    <SideNav.Item id="home" title="Home" />
    <SideNav.Item id="cart" title="Cart" />
    <SideNav.Item id="info" isDisabled title="Disabled" />
  </SideNav>
</div>
```

Set ```onNavClick``` on an individual ```SideNav.Item```.
```
<div style={{width: '30%'}}>
  <SideNav defaultSelected="home" altStyle>
    <SideNav.Item id="home" onNavClick={ () => (alert('Time to go Home!'))} title="Home" />
    <SideNav.Item id="cart" onNavClick={ () => (alert('You clicked the Cart!'))} title="Cart" />
    <SideNav.Item id="info" isDisabled title="Disabled" />
  </SideNav>
</div>
```

Manage selected item state manually using ```selected``` instead of ```defaultSelected```.
```
<div style={{width: '30%'}}>
  <SideNav selected={state.selected} onItemSelected={ (id) => {setState({selected: id})} }>
    <SideNav.Item id="home" title="Home" />
    <SideNav.Item id="cart" title="Cart" />
    <SideNav.Item id="help" title="Help" />
  </SideNav>
</div>
```
