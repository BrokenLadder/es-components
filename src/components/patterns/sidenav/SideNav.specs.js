/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import SideNav from './SideNav';

describe('drawer', () => {
  let instanceToRender;
  const onItemSelected = jest.fn();
  const onNavClick = jest.fn();

  beforeEach(() => {
    onItemSelected.mockClear();
    onNavClick.mockClear();

    instanceToRender = (
      <SideNav
        defaultSelected="home"
        onItemSelected={navId => {
          onItemSelected(navId);
        }}
      >
        <SideNav.Item
          id="home"
          className="home"
          onNavClick={navId => onNavClick(navId)}
        >
          Home
        </SideNav.Item>
        <SideNav.Item id="cart" className="cart">Cart</SideNav.Item>
        <SideNav.Item
          id="info"
          className="info"
          onNavClick={() => onNavClick()}
          isDisabled
        >
          Disabled
        </SideNav.Item>
      </SideNav>
    );
  });

  it('renders as expected', () => {
    const tree = renderer.create(instanceToRender).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('sets the selected state', () => {
    const navInstance = mount(instanceToRender);
    expect(navInstance.state('selected')).toBe('home');

    const navItem = navInstance.find('.cart');
    navItem.simulate('click');

    expect(navInstance.state('selected')).toBe('cart');
  });

  it('executes onItemSelected with the id of the nav item clicked', () => {
    const navInstance = mount(instanceToRender);
    const navItem = navInstance.find('.cart');

    navItem.simulate('click');
    expect(onItemSelected).toBeCalledWith('cart');
  });

  it('executes onNavClick when nav item clicked', () => {
    const navInstance = mount(instanceToRender);
    const navItem = navInstance.find('.home');

    navItem.simulate('click');
    expect(onNavClick).toBeCalledWith('home');
  });

  it('disabled item prevents onclick functions', () => {
    const navInstance = mount(instanceToRender);
    const navItem = navInstance.find('.info');

    navItem.simulate('click');
    expect(onNavClick).not.toBeCalled();
    expect(onItemSelected).not.toBeCalled();
  });
});
