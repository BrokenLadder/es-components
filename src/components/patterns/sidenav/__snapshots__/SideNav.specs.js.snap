exports[`drawer renders as expected 1`] = `
<nav
  className="sc-ifAKCX hvutRM">
  <ul>
    <li>
      <div
        className="home sc-bwzfXH YgUoe"
        onClick={[Function]}>
        <span
          className="sc-bxivhb khxFsg">
          Home
        </span>
        <i
          aria-hidden={true}
          className="sc-bdVaJa izWgQO"
          content="620"
          fontFamily="oecom-icons"
          fontSize="inherit" />
      </div>
    </li>
    <li>
      <div
        className="cart sc-bwzfXH jNiTgJ"
        onClick={[Function]}>
        <span
          className="sc-bxivhb khxFsg">
          Cart
        </span>
        <i
          aria-hidden={true}
          className="sc-bdVaJa izWgQO"
          content="620"
          fontFamily="oecom-icons"
          fontSize="inherit" />
      </div>
    </li>
    <li>
      <div
        className="info sc-bwzfXH bGNvpM"
        onClick={[Function]}>
        <span
          className="sc-bxivhb khxFsg">
          Disabled
        </span>
        <i
          aria-hidden={true}
          className="sc-bdVaJa izWgQO"
          content="620"
          fontFamily="oecom-icons"
          fontSize="inherit" />
      </div>
    </li>
  </ul>
</nav>
`;
